Sources/Algorithms/CommonPrefix.swift

```diff
@@ -0,0 +1,350 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public struct CommonPrefix<Base: Sequence, Other: Sequence> {
+  @usableFromInline
+  internal let base: Base
+  
+  @usableFromInline
+  internal let other: Other
+  
+  @usableFromInline
+  internal let areEquivalent: (Base.Element, Other.Element) -> Bool
+  
+  @inlinable
+  internal init(
+    base: Base,
+    other: Other,
+    areEquivalent: @escaping (Base.Element, Other.Element) -> Bool
+  ) {
+    self.base = base
+    self.other = other
+    self.areEquivalent = areEquivalent
+  }
+}
+
+extension CommonPrefix: Sequence {
+  public struct Iterator: IteratorProtocol {
+    @usableFromInline
+    internal var base: Base.Iterator
+    
+    @usableFromInline
+    internal var other: Other.Iterator
+    
+    @usableFromInline
+    internal let areEquivalent: (Base.Element, Other.Element) -> Bool
+    
+    @inlinable
+    internal init(
+      base: Base.Iterator,
+      other: Other.Iterator,
+      areEquivalent: @escaping (Base.Element, Other.Element) -> Bool
+    ) {
+      self.base = base
+      self.other = other
+      self.areEquivalent = areEquivalent
+    }
+    
+    public mutating func next() -> Base.Element? {
+      if let next = base.next(),
+         let otherNext = other.next(),
+         areEquivalent(next, otherNext) {
+        return next
+      } else {
+        return nil
+      }
+    }
```

Comment:
I think the iterator needs to add a `done` flag to its state.  For instance, if we do `"1234A56"` vs. `"1234B56"`, and we iterate past the first `nil`, we'll get a `"5"`, instead of `nil` forevermore.